flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java

```diff
@@ -114,6 +114,7 @@ public MailboxProcessor(
 		this.actionExecutor = Preconditions.checkNotNull(actionExecutor);
 		this.mailbox = Preconditions.checkNotNull(mailbox);
 		this.mainMailboxExecutor = Preconditions.checkNotNull(mainMailboxExecutor);
+		((MailboxExecutorImpl) this.mainMailboxExecutor).setMailboxProcessor(this);
```

Comment:
Couple of problems:
1. it changes an object passed from the outside
2. requires cast (could be solved by changing the signature of the constructor)

I'm thinking that maybe a better solution is to get rid of the `mainMailboxExecutor` and replace getter `getMainMailboxExecutor` with:
```
	public MailboxExecutor getMainMailboxExecutor() {
		return new MailboxExecutorImpl(mailbox, MIN_PRIORITY, actionExecutor, this);
	}
```
+
cache access to `mailboxProcessor.getMainMailboxExecutor()` inside `StreamTask` by creating a field in ` StreamTask`: `private final MailboxExecutor mainMailboxExecutor`?

This would solve the problem of circular dependency and nullable non final field inside `MailboxExecutorImpl`.

WDYT?